Feat/semantic-similarity-doc (#2261)
Kong/developer.konghq.com: app/ai-gateway/semantic-similarity.md

    "model": "text-embedding-3-large",
    "dimensions": 20
}
```

Creates the following embedding:

```json
{
	"object": "list",
	"data": [
		{
			"object": "embedding",
			"index": 0,
			"embedding": [
				0.26458353,
				-0.062855035,
				-0.14282244,
				0.18218088,
				-0.41043353,
				0.3704169,
				0.1712553,
				-0.10945333,
				-0.00060006406,
				0.10076551,
				-0.0697658,
				0.1779686,
				-0.3464596,
				0.028745485,
				0.3017042,
				0.2543161,
				-0.20916577,
				-0.06255886,
				-0.21469438,
				0.32934725
			]
		}
	],
	"model": "text-embedding-3-large",
	"usage": {
		"prompt_tokens": 28,
		"total_tokens": 28
	}
}
```

The `embedding` array contains 20 floating-point numbers—each one representing a dimension in the vector space.

{:.info}
> For simplicity, this example uses a reduced dimensionality of 20, though production models typically use `1536` or more.

### Practical considerations

If you use embedding models that support defining the dimensionality of the embedding output, you should consider how to balance accuracy and performance based on your use case.

> - **Lower dimensionality** improves speed and performance but might miss important nuance and lead to less accurate matches. This works well for simpler tasks like basic keyword matching or simple images, where hundreds of dimensions may suffice.
> 
> -  **Higher dimensionality** improves the granularity and nuance of similarity searches but increases storage and computation costs. This is needed for complex tasks like semantic text understanding or detailed images, where thousands of dimensions are often required.
> 
> However, dimensionality extremes at the far ends of the spectrum present significant drawbacks:

Reviewer: ```suggestion
However, dimensionality extremes at the far ends of the spectrum present significant drawbacks:
```

> {% table %}
> columns:
>   - title: Dimensionality range
>     key: range
>   - title: Drawbacks
>     key: drawbacks
> rows:
>   - range: "Very low (2–10 dimensions)"
>     drawbacks: "Fast but too simplistic; similarity search loses meaning. Like calling a movie simply 'good' or 'bad.'"
>   - range: "Very high (10,000+ dimensions)"
>     drawbacks: "Detailed but slow and expensive; can suffer from the 'curse of dimensionality,' where differences become less meaningful."
> {% endtable %}

Reviewer: ```suggestion
{% table %}
columns:
  - title: Dimensionality range
    key: range
  - title: Benefits
    key: benefits
  - title: Drawbacks
    key: drawbacks
rows:
  - range: "Lower dimensionality (2–10 dimensions)"
    benefits: |
      * Improves speed and performance
      * Works well for simpler tasks like basic keyword matching or simple images, where hundreds of dimensions may suffice.
    drawbacks: |
      * Can be too simplistic, like calling a movie simply "good" or "bad"
      * Might miss important nuance and lead to less accurate matches
  - range: "Higher dimensionality (10,000+ dimensions)"
    benefits: |
      * Improves the granularity and nuance of similarity searches
      * Useful for complex tasks like semantic text understanding or detailed images, where thousands of dimensions are often required.
    drawbacks: |
      * Increases storage and computation costs
      * Can suffer from the "curse of dimensionality", where differences become less meaningful.
{% endtable %}
```
Combining the content from the bullets with the table since the content was fairly similar. Check my formatting though

{:.success}
> Use moderate dimensionality when possible, and tune it based on both the complexity of your data and the responsiveness required by your application.

### Cosine and Euclidean similarity

> Kong AI Gateway supports both **cosine similarity** and **Euclidean distance** for vector comparisons, allowing you to choose the method best suited for your use case. You can configure the method using `config.vectordb.distance_metric` setting in the respective plugin.

Reviewer: ```suggestion
Kong AI Gateway supports both cosine similarity and Euclidean distance for vector comparisons, allowing you to choose the method best suited for your use case. You can configure the method using `config.vectordb.distance_metric` setting in the respective plugin.
```
